sys/net/gnrc/link_layer/netdev2/gnrc_netdev2_eth.c

```diff
@@ -0,0 +1,195 @@
+#include "net/gnrc.h"
+#include "net/gnrc/gnrc_netdev2.h"
+#include "net/ethernet/hdr.h"
+
+#include "od.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+static gnrc_pktsnip_t *_recv(gnrc_netdev2_t *gnrc_netdev2)
```

Comment:
But in general this should work (without error checking to make the code look a little simpler):

```C
if (bytes_expected) {
    ethernet_hdr_t *hdr;
    gnrc_pktsnip_t *netif_hdr, *eth_hdr;
    gnrc_pktsnip_t *pkt = gnrc_pktbuf_add(NULL, NULL, bytes_expected, GNRC_NETTYPE_UNDEF);
    int nread = dev->driver->recv(dev, pkt->data, bytes_expected);
    size_t data_len = (nread - sizeof(ethernet_hdr_t));
    if (nread < bytes_expected) {
         gnrc_pktbuf_realloc_data(pkt, nread);
    }
    eth_hdr = gnrc_pktbuf_mark(pkt, sizeof(ethernet_hdr_t), GNRC_NETTYPE_UNDEF);
    hdr = eth_hdr->data;
    gnrc_netif_hdr_init(netif_hdr->data, ETHERNET_ADDR_LEN, ETHERNET_ADDR_LEN);
    gnrc_netif_hdr_set_src_addr(netif_hdr->data, hdr->src, ETHERNET_ADDR_LEN);
    gnrc_netif_hdr_set_dst_addr(netif_hdr->data, hdr->dst, ETHERNET_ADDR_LEN);
    ((gnrc_netif_hdr_t *)netif_hdr->data)->if_pid = thread_getpid();
    gnrc_pktbuf_remove_snip(pkt, hdr);
    LL_APPEND(pkt, hdr);
    return pkt;
}
return NULL;
```
?